pandas/parser.pyx

```diff
@@ -1468,6 +1505,97 @@ cdef _string_box_decode(parser_t *parser, int col,
 
     return result, na_count
 
+@cython.boundscheck(False)
+cdef _categorical_convert(parser_t *parser, int col,
```

Comment:
The hash tables make this hard to re-use, the string functions have a table mapping bytes->object, where I need bytes->int here.  I could kill a lot of duplicated code with some templating, ala #13399, but I'm not sure that actually reduces complexity, because then there's another codegen step.